File: lua/vimade/util/matchers.lua
local M = {}

M.match_contains_any = function(target, value)
  -- second is the target and value a table?
  -- If yes, continue the recursion step
  if type(target) == 'table' and type(value) == 'table' then
    return M.match_table_any(target, value)
  -- otherwise outer compare the target to a real value
  -- from here on out
  elseif type(target) == 'table' then
    for i, v in pairs(target) do
      if M.match_contains_any(v, value) then
        return true
      end
    end
  else
    return M.match_primitive(target, value)
  end
end

M.match_table_any = function(target, value)
  for target_key, target_value in pairs(target) do
    if type(target_key) == 'number' then
      -- numbers are skipped as this is listy
      -- we assume things can be out of order here, so every
      -- item must be checked
      for real_key, real_value in pairs(value) do
        if M.match_contains_any(target_value, real_value) then
          return true
        end
      end
    else
      if M.match_contains_any(target_value, value[target_key], M.match_table_any) then
        return true
      end
    end
  end
  return false
end

M.each_match_contains_any = function(target, values)
 local count_ln = table.getn(values)
 local counts = {}
 for i, value in pairs(values) do
   if M._each_match_contains_any(target, value, counts, count_ln, i) == true then
     return true
   end
 end
end

M._each_match_contains_any = function(target, value, counts, counts_ln, i)
  for key, t in pairs(target) do
    --print(key, t)
    counts[key] = counts[key] or {count=0}
    local found = false
    if type(key) == 'number' then
      if type(t) == 'function' and type(value) == 'table' then
        for j, v in pairs(value) do
          if t(v) == true then
            found = true
          end
        end
      -- in this scenario every value key needs to be checked
      -- this is because we don't necessarily care about the order
      -- of a list and should accept any value present as a possible match
      -- consider target={{1}} and value={{2,1} f.  Even though 2 is index=0,
      -- this should still return true
      elseif type(t) == 'table' and type(value) == 'table' then
        for j, v in pairs(value) do
          if M._each_match_contains_any(t, v, counts[key], counts_ln, i) then
            -- finding a nested match means we are just done
            return true
          end
        end
      -- consider target={1} value={1}
      -- key=1, t=1, value is a table
      -- we need to see if t is present in value
      elseif type(value) == 'table' then
       for j, v in pairs(value) do
         if M.match_primitive(t, v) then
           found = true
         end
       end
      -- consider target={1} value=1
      -- key=1, t=1, t == 1
      elseif M.match_primitive(t, value) then
        found = true
      end
    else
      if type(t) == 'function' and type(value) == 'table' then
        if t(value[key]) == true then
          found = true
        end
      -- consider target={wtf={abc=1}} value={{wtf={abc=1}}}
      -- key=wtf, value={abc=1}  compare value[wtf]
      elseif type(t) == 'table' and type(value) == 'table' then
        if M._each_match_contains_any(t, value[key], counts[key], counts_ln, i) then
          return true
        end
      -- consider target={wtf=1} value={wtf=1}
      -- key=wtf, t=1, value={wtf=1}
      elseif type(value) == 'table' then
        --print('here', value[key], value)
        if M.match_primitive(t, value[key]) then
          found = true
        end
      elseif M.match_primitive(t, value) then
        found = true
      end
    end
    if found and counts[key].count == i - 1 then
      counts[key].count = i
    end
    if counts[key].count == counts_ln then
      return true
    end
  end
  return false
end

M.match_contains_all = function(target, value)
  return M.each_match_contains_all(target, {value})
end

M.match_table_all = function(target, value)
  for target_key, target_value in pairs(target) do
    local found = false
    if type(target_key) == 'number' then
      -- numbers are skipped as this is listy
      -- we assume things can be out of order here, so every
      -- item must be checked
      for real_key, real_value in pairs(value) do
        if type(target_value) == 'function' then
          found = target_value(real_value)
        elseif type(target_value) == 'table' and type(real_value) == 'table' then
          found = M.match_table_all(target_value, real_value)
        else
          -- outer compare
          found = M.match_contains_any(target_value, real_value)
        end
      end
    else
      if type(target_value) == 'function' then
        found = target_value(value[target_key])
      elseif type(target_value) == 'table' and type(value[target_key]) == 'table' then
        found = M.match_table_all(target_value, value[target_key])
      else
        found = M.match_contains(target_value, value[target_key])
      end
    end
    -- we don't care if the value has other keys, this is left match
    if not found then
      return false
    end
  end
  return true
end

M.each_match_contains_all = function(target, values)
  -- every target must contain every value
  for i, value in pairs(values) do
    local found
    -- second is the target and value a table?
    -- If yes, continue the recursion step
    if type(target) == 'table' and type(value) == 'table' then
      found = M.match_table_all(target, value)
    -- otherwise outer compare the target to a real value
    -- from here on out
    elseif type(target) == 'table' then
      for i, v in pairs(target) do
        if not M.match_contains_all(v, value) then
          found = false
          break
        end
      end
    else
      found = M.match_primitive(target, value)
    end
    if found == false then
      return false
    end
  end
end

M.match_primitive = function(target, value)
  if type(target) == 'function' then
    return target(value)
  elseif type(target) == 'boolean' then
    if target then
      return M.match_truthy(value)
    else
      return M.match_falsy(value)
    end
  elseif (type(target) == 'number' or type(target) == 'string')
    and (type(value) == 'number' or type(value) == 'string') then
    return target .. '' == value .. ''
  else
    return target == value
  end
end

M.match_truthy = function(value)
  return (value ~= nil and value ~= '' and value ~= false and value ~= 0) and true or false
end

M.match_falsy = function(value)
  return value ~= nil and (value == '' or value == false or value == 0) and true or false
end


M.match_contains_string = function(target, value)
  if target == nil then
    return false
  -- same as above function takes+riority
  elseif type(target) == 'function' then
    return target(value)
  -- target is a string, we do the comparison
  elseif type(target) == 'string'  then
    return M.match_string(target, value)
  -- target is a table, we need to see if value is within the table
  elseif type(target) == 'table' then
    for i, v in pairs(target) do
      if M.match_contains_string(v, value) then
        return true
      end
    end
  end
  return false
end

M.match_string = function(target, value)
  return string.find(string.lower(value), string.lower(target)) ~= nil
end

M.ContainsString = function(target)
  if type(target) == 'table' then
    return function(value)
      return M.match_contains_string(target, value)
    end
  else
    return function(value)
      return M.match_string(target, value)
    end
  end
end

M.ContainsAll = function(target)
  return function(value)
    return M.match_contains_all(target, value)
  end
end

M.EachContainsAll = function(target)
  return function(value)
    return M.each_match_contains_all(target, value)
  end
end

M.ContainsAny = function(target)
  return function(value)
    return M.match_contains_any(target, value)
  end
end

M.EachContainsAny = function(target)
  return function(values)
    return M.each_match_contains_any(target, values)
  end
end

M.IsFalsy = function ()
  return function (value)
    return M.match_falsy(value)
  end
end

M.IsTruthy = function ()
  return function (value)
    return M.match_truthy(value)
  end
end

M.StringMatcher = function (target)
  return function (value)
    if type(target) == 'table' then
      return M.match_contains_string(target, value)
    elseif type(target) == 'string' then
      return M.match_string(target, value)
    end
    return false
  end
end

return M

File: lib/vimade/util/matchers.py
import sys
import vim
M = sys.modules[__name__]

IS_V3 = False
if (sys.version_info > (3, 0)):
    IS_V3 = True
TYPE_VARS = type(vim.vars) # handle vim and nvim remote objects
TYPE_OPTS = type(vim.options) # handle vim and nvim remote objects
DICTIONARY_LIKE_TYPES = (list, dict, TYPE_VARS, TYPE_OPTS)
def _bytes_str_v3(value):
  if type(value) == bytes:
    return str(value, 'utf-8')
  return value
def _bytes_str_v2(value):
  if type(value) == bytes:
    return str(value)
  return value
SAFE_BYTES_STR = _bytes_str_v3 if IS_V3 else _bytes_str_v2

def _safe_get(value, key):
  if type(value) == list:
    if type(key) == int and key < len(value):
      return value[key]
  elif type(value) == dict:
    return value.get(key)
  elif type(value) == TYPE_VARS:
    return value.get(key)
  elif type(value) == TYPE_OPTS:
    if key in value:
      return value[key]
  return None

def _pairs(input):
  if type(input) == list:
    return enumerate(input)
  elif type(input) == dict:
    return input.items()

def match_contains_any(target, value):
  if type(target) in DICTIONARY_LIKE_TYPES and (type(value) in DICTIONARY_LIKE_TYPES):
    return M.match_table_any(target, value)
  elif type(target) in DICTIONARY_LIKE_TYPES:
    for i, v in _pairs(target):
      if M.match_contains_any(v, value):
        return True
  else:
    return M.match_primitive(target, value)

def match_table_any(target, value):
  for target_key, target_value in _pairs(target):
    target_key = SAFE_BYTES_STR(target_key)
    if type(target_key) == int:
      for real_key, real_value in _pairs(value):
        real_key = SAFE_BYTES_STR(real_key)
        if M.match_contains_any(target_value, real_value):
          return True
    else:
      if M.match_contains_any(target_value, _safe_get(value, target_key)):
        return True
  return False
  

def each_match_contains_any(target, values):
  count_ln = len(values)
  counts = {}
  for i, value in _pairs(values):
    if M._each_match_contains_any(target, value, counts, count_ln, i) == True:
      return True

def _each_match_contains_any(target, value, counts, counts_ln, i):
  for key, t in _pairs(target):
    key = SAFE_BYTES_STR(key)
    counts[key] = counts.get(key, {'count': 0})
    found = False
    if type(key) == int:
      if callable(t) and type(value) in DICTIONARY_LIKE_TYPES:
        for j, v in _pairs(value):
          if t(v) == True:
            found = True
      elif type(t) in DICTIONARY_LIKE_TYPES and type(value) in DICTIONARY_LIKE_TYPES:
        for j, v in _pairs(value):
          if M._each_match_contains_any(t, v, counts[key], counts_ln, i):
            return True
      elif type(value) in DICTIONARY_LIKE_TYPES:
        for j, v in _pairs(value):
          if M.match_primitive(t, v):
            found = True
      elif M.match_primitive(t, value):
        found = True
    else:
      if callable(t) and type(value) in DICTIONARY_LIKE_TYPES:
        if t(value[key]) == True:
          found = True
      elif type(t) in DICTIONARY_LIKE_TYPES and type(value) in DICTIONARY_LIKE_TYPES:
        if M._each_match_contains_any(t, _safe_get(value, key), counts[key], counts_ln, i):
          found = True
      elif type(value) in DICTIONARY_LIKE_TYPES:
        if M.match_primitive(t, _safe_get(value, key)):
          found = True
      elif M.match_primitive(t, value):
        found = True
    if found and counts[key]['count'] == i:
      counts[key]['count'] = i + 1
    if counts[key]['count'] == counts_ln:
      return True
  return False


def match_contains_all(target, value):
  return M.each_match_contains_all(target, [value])

def match_table_all(target, value):
  for target_key, target_value in _pairs(target):
    target_key = SAFE_BYTES_STR(target_key)
    found = False
    if type(target_key) == int:
      # numbers are skipped as this is listy
      # we assume things can be out of order here, so every
      # item must be checked
      for real_key, real_value in _pairs(value):
        real_key = SAFE_BYTES_STR(real_key)
        if callable(target_value):
          found = target_value(real_value)
        elif type(target_value) in DICTIONARY_LIKE_TYPES and type(real_value) in DICTIONARY_LIKE_TYPES:
          found = M.match_table_all(target_value, real_value)
        else:
          found = M.match_contains_any(target_value, real_value)
    else:
      if callable(target_value):
        found = target_value(_safe_get(value, target_key))
      elif type(target_value) in DICTIONARY_LIKE_TYPES and type(_safe_get(value, target_key)) in DICTIONARY_LIKE_TYPES:
        found = M.match_table_all(target_value, _safe_get(value, target_key))
      else:
        # outer compare
        found = M.match_contains(target_value, _safe_get(value, target_key))
    # we don't care if the value has other keys, this is left match
    if not found:
      return False
  return True

def each_match_contains_all(target, values):
  # every target must contain every value
  for i, value in _pairs(values):
    found = None
    # second is the target and value a table?
    # If yes, continue the recursion step
    if type(target) in DICTIONARY_LIKE_TYPES and type(value) in DICTIONARY_LIKE_TYPES:
      found = M.match_table_all(target, value)
    # otherwise outer compare the target to a real value
    # from here on out
    elif type(target) in DICTIONARY_LIKE_TYPES:
      for i, v in _pairs(target):
        if not M.match_contains_all(v, value):
          found = false
          break
    else:
      found = M.match_primitive(target, value)
    if found == False:
      return False

def match_primitive(target, value):
  target = SAFE_BYTES_STR(target)
  value = SAFE_BYTES_STR(value)

  if callable(target):
    return target(value)
  elif type(target) == bool:
    if target:
      return M.match_truthy(value)
    else:
      return M.match_falsy(value)
  elif (type(target) in (int, float, str)) and (type(value) in (int, float, str)):
    return str(target) + '' == str(value) + ''
  else:
    return target == value

def match_truthy(value):
  return (value != None and value != '' and value != False and value != 0) and True or False

def match_falsy(value):
  return value != None and (value == '' or value == False or value == 0) and True or False

def match_contains_string(target, value):
  if target == None:
    return False
  # custom function
  elif callable(target):
    return target(value)
  # target is a string, we do the comparison
  elif type(target) in (str, bytes):
    return M.match_string(target, value)
  # target is a table, we need to see if value is within the table
  elif type(target) in (list, dict):
    for i,v in _pairs(target):
      if M.match_contains_string(v, value):
        return True
  return False

def match_string(target, value):
  target = SAFE_BYTES_STR(target)
  value = SAFE_BYTES_STR(value)
  if type(target) == str:
    return value != '' and ((value+'').lower() in (target+'').lower())
  return target == True and value != ''

def ContainsString(target):
  if type(target) in (list, dict):
    def _ContainsString(value):
      return M.match_contains_string(target, value)
    return _ContainsString
  else:
    def _ContainsString(value):
      return M.match_string(target, value)
    return _ContainsString

def ContainsAll(target):
  def _ContainsAll(value):
    return M.match_contains_all(target, value)
  return _ContainsAll

def EachContainsAll(target):
  def _EachContainsAll(value):
    return M.each_match_contains_all(target, value)
  return _EachContainsAll

def ContainsAny(target):
  def _ContainsAny(value):
    return M.match_contains_any(target, value)
  return _ContainsAny

def EachContainsAny(target):
  def _EachContainsAny(values):
    return M.each_match_contains_any(target, values)
  return _EachContainsAny

def IsFalsy(target):
  def _IsFalsy(value):
    return M.match_falsy(value)
  return _IsFalsy

def IsTruthy(target):
  def _IsTruthy(value):
    return M.match_truthy(value)
  return _IsTruthy

def StringMatcher(target):
  def _StringMatcher(value):
    if type(target) in DICTIONARY_LIKE_TYPES:
      return M.match_contains_string(target, value)
    elif type(target) in (str, bytes):
      return M.match_string(target, value)
    return False
  return _StringMatcher
